tsdb/index/index.go

```diff
@@ -1424,10 +1434,22 @@ func (r *Reader) Close() error {
 }
 
 func (r *Reader) lookupSymbol(o uint32) (string, error) {
+	cacheIndex := o % valueSymbolsCacheSize
+	if cached := r.valueSymbols[cacheIndex]; cached.index == o && cached.symbol != "" {
+		return cached.symbol, nil
+	}
+
 	if s, ok := r.nameSymbols[o]; ok {
 		return s, nil
 	}
-	return r.symbols.Lookup(o)
+
+	s, err := r.symbols.Lookup(o)
+	if err != nil {
+		return s, err
+	}
+	r.valueSymbols[cacheIndex].index = o
```

Comment:
Can't `lookupSymbol()` be called concurrently? If yes, then there's a race condition on `valueSymbols`.

Reply:
Correct, for the query path, and putting a lock around it negates most of the gains. For the Cortex type use cases it might be worthwhile to use a cache like this for the compaction path only, but I am not sure the extra complexity is worthwhile.